File: lib/statuses.ex
defmodule Statuses do
  @moduledoc """
  The `Statuses` module create a list of
  statuses to be used in Elixir projects
  """

  @doc """
    `parse_json` returns a list of maps containing status information. <br />

    The format returned is:
    ```elixir
    [
      %Status{code: 4, desc: "Items marked as done are complete", text: "done"},
      %Status{
        code: 5,
        desc: "A flagged App, Item or Person requires admin attention",
        text: "flagged"
      },
      %{...},
      ...
    ]
    ```
  """
  defp parse_json do
    {:ok, cwd} = File.cwd

    # we need this `cd` to locate the file in `/deps`
    case cwd =~ "/statuses" do

      true ->
        File.read!("statuses.json")
        |> Jason.decode!()

      # coveralls-ignore-start

      false -> # temporarily `cd` into `deps/quotes` dir and read `quotes.json` file:
        File.cd!("deps/quotes")
        data = File.read!("quotes.json")
        |> Jason.decode!()
        # change back into the root directory
        File.cd!("../..")
        data # return the decoded (parsed) JSON data

      # coveralls-ignore-stop
    end
  end


  @doc """
  Returns list of the statuses.
  """
  @spec get_statuses() :: list(Status.t())
  def get_statuses do
    parse_json()
    |> Enum.map(fn obj -> struct(Status, [code: obj["code"], desc: obj["desc"], text: String.to_atom(obj["text"])]) end)
  end
end

File: lib/statuses.dart
library statuses;

import 'dart:convert';
import 'dart:io';

/// Class holding the status information
class StatusInfo {
  final int code;
  final String desc;
  final String text;

  StatusInfo(this.code, this.desc, this.text);

  factory StatusInfo.fromJson(Map<String, dynamic> json) {
    return StatusInfo(
      json['code'],
      json['desc'],
      json['text']
    );
  }
}

/// Static class that can be invoked to fetch the statuses
class Statuses {

  /// Loads the information from the `.json` file containing the status list.
  static Future<List<StatusInfo>> _statusesList() async {
    String jsonString = await File('statuses.json').readAsString();
    List<dynamic> jsonList = json.decode(jsonString);
    List<StatusInfo> statuses = jsonList.map((json) => StatusInfo.fromJson(json)).toList();
    return statuses;
  }

  /// Returns a list of statuses
  static Future<List<StatusInfo>> list() async {
    return await _statusesList();
  }
}